torchbiggraph/storage_repository.py

```diff
@@ -0,0 +1,255 @@
+            self.put_to_hdfs()
+
+        return True
+
+    @staticmethod
+    def get_hdfs_path(path) -> str:
+        return str(path).replace(str(Path.cwd()), '', 1)
+
+    @staticmethod
+    def hdfs_file_exists(path: str) -> bool:
+        rcode = 1
+        try:
+            rcode, output, errors = run_external_cmd(Constants.TEST_FILE.format(remote_path=path))
+        except:
+            pass
+
+        return rcode == 0
+
+    def get_from_hdfs(self, reload):
+        local_path = self._path
+        hdfs_loc = self.get_hdfs_path(local_path)
+
+        if hdfs_loc != str(local_path):
+            # check if hdfs file exists before running get command
+            if self.hdfs_file_exists(hdfs_loc):
+                local_path.parent.mkdir(parents=True, exist_ok=True)
+
+                if local_path.exists():
+                    if reload:
+                        local_path.unlink()
+                        run_external_cmd(Constants.GET.format(remote_path=hdfs_loc, local_path=str(local_path)))
+                    else:
+                        logger.info(f"Skip get: reload is {reload} and local file exists : {local_path}")
+                else:
+                    run_external_cmd(Constants.GET.format(remote_path=hdfs_loc, local_path=str(local_path)))
+            else:
+                raise FileNotFoundError(f"File {hdfs_loc} not found.")
+        else:
+            logger.info(f"identical local and hdfs path. Skipping get {str(local_path)}")
+
+    def put_to_hdfs(self):
+        local_loc_str = str(self._path)
+        hdfs_loc = self.get_hdfs_path(local_loc_str)
+
+        if self._path.exists():
+            if hdfs_loc != local_loc_str:
+                run_external_cmd(Constants.PUT.format(local_path=local_loc_str, remote_path=hdfs_loc))
+            else:
+                logger.info('identical local and hdfs path. Skipping put ..', local_loc_str)
+        else:
+            raise FileNotFoundError("File " + local_loc_str + " not found.")
+
+
+class AbstractPath(type(pathlib.Path())):
+    @abstractmethod
+    def __init__(self, path: Union[str, Path]):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __truediv__(self, key: str) -> 'AbstractPath':
+        raise NotImplementedError
+
+    @abstractmethod
+    def __str__(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def parent(self) -> 'AbstractPath':
+        raise NotImplementedError
+
+    @abstractmethod
+    def resolve(self, strict: bool = False) -> 'AbstractPath':
+        raise NotImplementedError
+
+    @abstractmethod
+    def open(self, mode='r', buffering=-1, encoding=None,
+             errors=None, newline=None, **kwargs):
+        pass
+
+    @abstractmethod
+    def cleardir(self) -> None:
+        raise NotImplementedError
+
+CUSTOM_PATH = PluginRegistry[AbstractPath]()
+
+@CUSTOM_PATH.register_as("")
+@CUSTOM_PATH.register_as("file")
+class LocalPath(AbstractPath):
```

Comment:
If I understand correctly, all these wrappers are due to the fact that you're trying to reuse the same `Storage` classes for File and HDFS. We already created a set of abstractions for different storages, please use those instead of creating a new indirection point.